### ISSUE TITLE
pg_dump errors out when COPYing out distributed tables 

### ISSUE THREAD

**Issue by lithp** [citusdata/citus#348]
In [this thread](https://groups.google.com/a/citusdata.com/d/topic/dev/2_j8LYkqbeA/discussion) it's mentioned that running pg_dump on the master returns a nice error message:

```
$ /opt/citusdb/4.0/bin/pg_dump postgres > citus_master_dump
pg_dump: [archiver (db)] query failed: ERROR:  cannot execute COPY on a distributed table on master node
```

But when I run it myself I get:

```
$ pg_dump -p 9700 -f wikipedia_edits
pg_dump: [sorter] WARNING: could not resolve dependency loop among these items:
pg_dump: [sorter]   TABLE wikipedia_edits  (ID 188 OID 16460)
pg_dump: [archiver (db)] query failed: ERROR:  cannot execute nextval() in a read-only transaction
pg_dump: [archiver (db)] query was: COPY public.wikipedia_edits (reported_time, ingest_time, source, bot, wiki, namespace, title, comment, minor, type, old_length, new_length) TO stdout;
```

**Comment by lfittl** [comment]
For what its worth, I've been seeing the same dependency loop warnings when testing with Rails:

  ```
  pg_dump: [sorter] WARNING: could not resolve dependency loop among these items:
  pg_dump: [sorter]   TABLE impressions  (ID 240 OID 19654)
  pg_dump: [sorter] WARNING: could not resolve dependency loop among these items:
  pg_dump: [sorter]   TABLE clicks  (ID 241 OID 19661)
  ```

(see #595)

**Comment by keithf4** [comment]
Seeing the same dependency loop error on PostgreSQL 9.5/CitusDB 5.2-1. This is just using the same setup as given in the citus documentation (I just renamed the table with a _hash suffx)

http://docs.citusdata.com/en/v5.1/dist_tables/hash_distribution.html

```
$ pg_dump -Fc -d keith -f testing.pgr
pg_dump: [sorter] WARNING: could not resolve dependency loop among these items:
pg_dump: [sorter]   TABLE github_events_hash  (ID 192 OID 17517)
pg_dump: [sorter] WARNING: could not resolve dependency loop among these items:
pg_dump: [sorter]   TABLE github_events_append  (ID 193 OID 22283)
pg_dump: [sorter] WARNING: could not resolve dependency loop among these items:
pg_dump: [sorter]   TABLE wikipedia_changes_hash  (ID 195 OID 23306)
pg_dump: [sorter] WARNING: could not resolve dependency loop among these items:
pg_dump: [sorter]   TABLE wikipedia_editors_hash  (ID 196 OID 23312)
pg_dump: [archiver (db)] query failed: ERROR:  cannot execute nextval() in a read-only transaction
pg_dump: [archiver (db)] query was: COPY public.github_events_append (event_id, event_type, event_public, repo_id, payload, repo, actor, org, created_at) TO stdout;
```

**Comment by anarazel** [comment]
On 2016-02-16 15:47:56 -0800, Brian Cloutier wrote:
>     pg_dump: [archiver (db)] query failed: ERROR:  cannot execute nextval() in a read-only transaction
>     pg_dump: [archiver (db)] query was: COPY public.wikipedia_edits (reported_time, ingest_time, source, bot, wiki, namespace, title, comment, minor, type, old_length, new_length) TO stdout;

This is suggesting a bug in the copy code afaics.

**Comment by ozgune** [comment]
@anarazel / @lithp -- Do we run ``nextval()`` as part of ``COPY .. TO STDOUT``?

**Comment by anarazel** [comment]
On 2016-08-19 10:35:46 -0700, Ozgun Erdogan wrote:
> @anarazel / @lithp -- Do we run ``nextval()`` as part of ``COPY .. TO STDOUT``?

We shouldn't, that'd be a bug.

**Comment by lithp** [comment]
The utility hook [messes with](https://github.com/citusdata/citus/blob/master/src/backend/distributed/executor/multi_utility.c#L448) COPY TO if it's run on a distributed table and makes sure it reads from a query instead of the (empty) master table. Then passes the copy statement right back to the existing COPY machinery. The existing machinery doesn't call nextval() so we don't.

**Comment by lithp** [comment]
I also reported this error before we added function evaluation :)

**Comment by anarazel** [comment]
The nextval originates in assigning job-ids for the remote queries:
```
#0  PreventCommandIfReadOnly (cmdname=0xac416a "nextval()") at /home/andres/src/postgresql-9.5/src/backend/tcop/utility.c:229
#1  0x0000000000624368 in nextval_internal (relid=16419) at /home/andres/src/postgresql-9.5/src/backend/commands/sequence.c:566
#2  0x0000000000624290 in nextval_oid (fcinfo=0x7ffc184557b0) at /home/andres/src/postgresql-9.5/src/backend/commands/sequence.c:530
#3  0x0000000000932357 in DirectFunctionCall1Coll (func=0x62426f <nextval_oid>, collation=0, arg1=16419)
    at /home/andres/src/postgresql-9.5/src/backend/utils/fmgr/fmgr.c:1027
#4  0x00007fb4e6d54517 in UniqueJobId () at /home/andres/src/citusdb/src/backend/distributed/planner/multi_physical_planner.c:1660
#5  0x00007fb4e6d5457a in BuildJob (jobQuery=0x19336f0, dependedJobList=0x0)
    at /home/andres/src/citusdb/src/backend/distributed/planner/multi_physical_planner.c:1673
#6  0x00007fb4e6d51fdc in BuildJobTree (multiTree=0x1931938) at /home/andres/src/citusdb/src/backend/distributed/planner/multi_physical_planner.c:390
#7  0x00007fb4e6d51a9a in MultiPhysicalPlanCreate (multiTree=0x1931938)
    at /home/andres/src/citusdb/src/backend/distributed/planner/multi_physical_planner.c:208
#8  0x00007fb4e6d5ad21 in CreatePhysicalPlan (originalQuery=0x195ec40, query=0x195ed58, restrictionContext=0x1966730)
    at /home/andres/src/citusdb/src/backend/distributed/planner/multi_planner.c:128
#9  0x00007fb4e6d5ac40 in multi_planner (parse=0x195ed58, cursorOptions=0, boundParams=0x0)
    at /home/andres/src/citusdb/src/backend/distributed/planner/multi_planner.c:78
#10 0x0000000000720fb8 in planner (parse=0x195ed58, cursorOptions=0, boundParams=0x0)
    at /home/andres/src/postgresql-9.5/src/backend/optimizer/plan/planner.c:156
#11 0x00000000005e7894 in BeginCopy (is_from=0 '\000', rel=0x0, raw_query=0x18db7d0, 
    queryString=0x18ec398 "COPY public.github_events (event_id, event_type, event_public, repo_id, payload, repo, actor, org, created_at) TO stdout;", 
    queryRelId=0, attnamelist=0x18dbb20, options=0x0) at /home/andres/src/postgresql-9.5/src/backend/commands/copy.c:1416
#12 0x00000000005e84a8 in BeginCopyTo (rel=0x0, query=0x18db7d0, 
    queryString=0x18ec398 "COPY public.github_events (event_id, event_type, event_public, repo_id, payload, repo, actor, org, created_at) TO stdout;", 
    queryRelId=0, filename=0x0, is_program=0 '\000', attnamelist=0x18dbb20, options=0x0) at /home/andres/src/postgresql-9.5/src/backend/commands/copy.c:1688
#13 0x00000000005e5f96 in DoCopy (stmt=0x191ae28, 
    queryString=0x18ec398 "COPY public.github_events (event_id, event_type, event_public, repo_id, payload, repo, actor, org, created_at) TO stdout;", 
    processed=0x7ffc18456268) at /home/andres/src/postgresql-9.5/src/backend/commands/copy.c:949
#14 0x00000000007eb611 in standard_ProcessUtility (parsetree=0x191ae28, 
    queryString=0x18ec398 "COPY public.github_events (event_id, event_type, event_public, repo_id, payload, repo, actor, org, created_at) TO stdout;", 
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, dest=0x18ed708, completionTag=0x7ffc18456640 "")
    at /home/andres/src/postgresql-9.5/src/backend/tcop/utility.c:541
---Type <return> to continue, or q <return> to quit---
#15 0x00007fb4e6d3830e in multi_ProcessUtility (parsetree=0x191ae28, 
    queryString=0x18ec398 "COPY public.github_events (event_id, event_type, event_public, repo_id, payload, repo, actor, org, created_at) TO stdout;", 
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, dest=0x18ed708, completionTag=0x7ffc18456640 "")
    at /home/andres/src/citusdb/src/backend/distributed/executor/multi_utility.c:303
#16 0x00000000007eb0d6 in ProcessUtility (parsetree=0x18ed360, 
    queryString=0x18ec398 "COPY public.github_events (event_id, event_type, event_public, repo_id, payload, repo, actor, org, created_at) TO stdout;", 
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, dest=0x18ed708, completionTag=0x7ffc18456640 "")
    at /home/andres/src/postgresql-9.5/src/backend/tcop/utility.c:330
#17 0x00000000007ea207 in PortalRunUtility (portal=0x186b978, utilityStmt=0x18ed360, isTopLevel=1 '\001', dest=0x18ed708, completionTag=0x7ffc18456640 "")
    at /home/andres/src/postgresql-9.5/src/backend/tcop/pquery.c:1183
#18 0x00000000007ea3d5 in PortalRunMulti (portal=0x186b978, isTopLevel=1 '\001', dest=0x18ed708, altdest=0x18ed708, completionTag=0x7ffc18456640 "")
    at /home/andres/src/postgresql-9.5/src/backend/tcop/pquery.c:1314
#19 0x00000000007e9986 in PortalRun (portal=0x186b978, count=9223372036854775807, isTopLevel=1 '\001', dest=0x18ed708, altdest=0x18ed708, 
    completionTag=0x7ffc18456640 "") at /home/andres/src/postgresql-9.5/src/backend/tcop/pquery.c:812
#20 0x00000000007e3b3d in exec_simple_query (
    query_string=0x18ec398 "COPY public.github_events (event_id, event_type, event_public, repo_id, payload, repo, actor, org, created_at) TO stdout;")
    at /home/andres/src/postgresql-9.5/src/backend/tcop/postgres.c:1104
#21 0x00000000007e7bf8 in PostgresMain (argc=1, argv=0x184d698, dbname=0x184d4f8 "postgres", username=0x184d4d8 "andres")
    at /home/andres/src/postgresql-9.5/src/backend/tcop/postgres.c:4045
#22 0x00000000007666c9 in BackendRun (port=0x1879b90) at /home/andres/src/postgresql-9.5/src/backend/postmaster/postmaster.c:4243
#23 0x0000000000765e2d in BackendStartup (port=0x1879b90) at /home/andres/src/postgresql-9.5/src/backend/postmaster/postmaster.c:3917
#24 0x0000000000762635 in ServerLoop () at /home/andres/src/postgresql-9.5/src/backend/postmaster/postmaster.c:1688
#25 0x0000000000761c6a in PostmasterMain (argc=17, argv=0x184c100) at /home/andres/src/postgresql-9.5/src/backend/postmaster/postmaster.c:1296
#26 0x00000000006ba8e0 in main (argc=17, argv=0x184c100) at /home/andres/src/postgresql-9.5/src/backend/main/main.c:228
(gdb) 
```

**Comment by lfittl** [comment]
FWIW, extracted the dependency loop problem into its own issue per our discussion today: #866

**Comment by anarazel** [comment]
For posterity's sake: "The pg_dump error message has become less useful?" isn't a great description of an issue - hard to understand what that could mean when just looking at the title.

**Comment by lithp** [comment]
@anarazel What do you propose as a better description? `pg_dump used to print "cannot execute COPY on a distributed table on master node" when erroring out but now prints something which gives users less guidance`?

You've changed the title but in the process also changed the issue! I didn't mean "we should support pg_dump", I literally meant "seeing as that we don't support pg_dump, we used to kindly inform the user that this intentionally doesn't work whereas we now error out with something the user might construe as being their fault".

**Comment by anarazel** [comment]
> @anarazel What do you propose as a better description?

There's a lot of possibilities. Something like "Make warning about failing COPY in pg_dump more descriptive, again" could do the trick.   Your previous title is a question, not an issue title. 

>  I literally meant "seeing as that we don't support pg_dump, we used to kindly inform the user that this intentionally doesn't work whereas we now error out with something the user might construe as being their fault".

Well, but that's not really the case. We previously didn't support COPY, now we just don't support COPY in readonly transactions.

**Comment by lfittl** [comment]
We should be able to close this once #1208 is merged.

**Comment by marcocitus** [comment]
Fixed by #1208

**Comment by ozgune** [comment]
@lfittl -- Quick question. I remember that we decided to split this issue into two issues -- one related to the warnings that we see related to schema-only dumps (#866) and another around taking pg_dumps (this issue).

Does #1208 fix this issue, #866, or both?

**Comment by anarazel** [comment]
@ozgune It only fixes #1208, it doesn't do anything about #866.

**Comment by lfittl** [comment]
Yup, what Andres said. The circular dependency issue (and related slow structural dump) is still there when I tested the branch (which makes sense, since its a different underlying problem).
